sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -808,8 +799,15 @@ def register_resource(
     # passed to.  However, those futures won't actually resolve until the RPC returns
     resolvers = rpc.transfer_properties(res, props)
 
-    async def do_register():
+    async def do_register() -> None:
         try:
+            from ..resource import (  # pylint: disable=import-outside-toplevel
+                ResourceOptions,
+            )
+
```

Comment:
Why do we need this?
We have an import of this above that's conditional on TYPE_CHECKING.
Can we just move it to top-level?

Reply:
No because that causes a circular import between runtime and the top level package. There's quite a few places through the SDK like this 😢

Reply:
I assumed that the import in function was to work around a cycle, but then does the conditional TYPE_CHECKING import not have the same problem? Is type checking analysis a different phase that doesn't actually run the import?

Reply:
>  Is type checking analysis a different phase that doesn't actually run the import?

Yeh the type checker can't actually do a normal import because that could run arbitrary python code, so it has it's own import system.